### PR TITLE
Add deny authorizer API

### PR DESCRIPTION
**Pull request by enj**
https://trello.com/c/uxG5LtTX

client pull: https://github.com/openshift/client-go/pull/53
origin pull: https://github.com/openshift/origin/pull/19475

### CODE REVIEW

File: authorization/v1alpha1/register.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	GroupName     = "authorization.openshift.io"
+	GroupVersion  = schema.GroupVersion{Group: GroupName, Version: "v1alpha1"}
```
Comment (deads2k) [inline authorization/v1alpha1/register.go]: in the group level install, set the prioritized versions to v1 then v1alpha1

File: authorization/v1alpha1/types.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// AccessRestrictionSpec holds the matching requirements.
+// MatchAttributes and DeniedSubjects (blacklist) are required.
+// AllowedSubjects (whitelist) may optionally be specified.
+type AccessRestrictionSpec struct {
+	// If these rules cover the current request, then this restriction applies.
+	// If AllowedSubjects is set, those subjects can perform the matching actions even if they exist in DeniedSubjects.
+	// The subjects in DeniedSubjects are restricted from performing the matching actions unless they exist in AllowedSubjects.
+	// If the subject does not exist in either AllowedSubjects or DeniedSubjects, the request is allowed.
```
Comment (deads2k) [inline authorization/v1alpha1/types.go]: ", then this authorizer has no opinion on this request.  "

**Comment by deads2k** [comment]
Minor comments.

@liggitt we reviewed the google doc.  you want a shot at this?

File: authorization/v1alpha1/register.go (hunk 1 above)
Comment (enj) [reply authorization/v1alpha1/register.go]: Fixed.

File: authorization/v1alpha1/types.go (hunk 2 above)
Comment (enj) [reply authorization/v1alpha1/types.go]: Updated.

**Comment by enj** [comment]
> @liggitt we reviewed the google doc. you want a shot at this?

Pretty sure Jordan hates the very idea of this API :smile:

File: authorization/v1alpha1/register.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	schemeBuilder = runtime.NewSchemeBuilder(addKnownTypes, corev1.AddToScheme, rbacv1.AddToScheme)
```
Comment (deads2k) [inline authorization/v1alpha1/register.go]: you don't appear to use corev1

File: authorization/v1alpha1/types.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	// Note that this only prevents a denial due to the access restriction.
+	// The subject must still have a matching RBAC binding to actually perform the current action.
+	// Optional.
+	AllowedSubjects []SubjectMatcher `json:"allowedSubjects,omitempty" protobuf:"bytes,2,opt,name=allowedSubjects"`
+
+	// The blacklist of subjects that are not allowed to perform the actions defined by MatchAttributes.
+	// AllowedSubjects has priority over this list.  It is ignored if the subject exists in AllowedSubjects.
+	// This restriction is processed before all RBAC data, and thus will reject actions that RBAC may otherwise permit.
+	// Required.
+	DeniedSubjects []SubjectMatcher `json:"deniedSubjects,omitempty" protobuf:"bytes,3,opt,name=deniedSubjects"`
```
Comment (deads2k) [inline authorization/v1alpha1/types.go]: not omitempty

**Comment by deads2k** [comment]
This isn't going to make 1.11.  Update the origin pull to get an lgtm and be green. See the detailed instructions here: https://github.com/openshift/api

**Pull request closed without merging**